Revamp building hours

mattk410 opened the pull request:
This adds a collapsible view of building hours.

Issue #143

![simulator screen shot sep 10 2016 11 51 34 am](https://cloud.githubusercontent.com/assets/6880028/18412148/5502b302-774d-11e6-9ded-ecfeb5a8223f.png)

Commit:
Added CollapsibleBlock component which adds collapsible capabilities to blocks of components.
Added a function to display all hours for a building inside its collapsible component.

Commit:
Merge branch 'master' into revamp-building-hours

Commit:
Added CollapsibleBlock component which adds collapsible capabilities to blocks of components.
Added a function to display all hours for a building inside its collapsible component.

Commit:
Merge branch 'master' into revamp-building-hours

hawkrives commented on views/building-hours/building.js:
```diff
@@ -16,75 +15,106 @@ import * as c from '../components/colors'
 import {formatBuildingHours} from './format-building-hours'
+import CollapsibleBlock from '../components/collapsibleBlock'
+import {allBuildingHours} from './all-building-hours.js'
+import Dimensions from 'Dimensions'
```
I think this also exists as an import from `"react-native"` itself; could you import from there, please?

hawkrives commented on views/building-hours/all-building-hours.js:
```diff
@@ -0,0 +1,29 @@
+  let hoursString = ''
+  for (let i = 0; i < 7; i++) {
+    let day = moment().add(i, 'days').format('dddd')
+    let d = moment().add(i, 'days').format('ddd')
```
Two notes here: one, could you create one moment and query it twice, instead of creating two moments? I have an irrational fear of the two moments getting created across a day divide so they don't match… 

The other note is, could you use moment-timezone set to … I think we use America/Winnipeg? Otherwise is used the device's timezone, which is fine but the rest of the building hours stuff uses the explicit America/Winnipeg zone. I think it'll be really confused when I try it in Japan if half of it's in JST and half is in CST…

hawkrives commented on views/building-hours/all-building-hours.js:
```diff
@@ -0,0 +1,29 @@
+
+export function allBuildingHours(info, style) {
+  let dayTimes = []
+  let hoursString = ''
```
Why are you defining `hoursString` all the way out here? Could you move it a bit closer to where it's used, or at least to inside the loop? Thx.

elijahverdoorn commented on views/components/collapsibleBlock.js:
```diff
@@ -0,0 +1,59 @@
+    isCollapsed: true,
+  }
+
+  style() {
```
This should be in a stylesheet, as it is in the rest of the project.

hawkrives replied:
The advantage to `StyleSheet.create` is that RN can process the stylesheet once, then reference a cache to load the styles in the future, rather than needing to build the entire style tree every time. Could you extract `this.style` into an external `styles` variable created with `StyleSheet.create`, then reference it in the `render` method, please?

That will require that you use the array form of `style={[styles.collapsedContent, {backgroundColor: this.props.borderColor}]}` for the `collapsedContent` block.

hawkrives commented on views/building-hours/building.js:
```diff
@@ -16,75 +15,106 @@ import * as c from '../components/colors'
+
+
+
+let style = function(){
```
Why did you make style a function? If it's so that `getContainerWidth` is updated when the screen moves, I'd recommend moving just _that_ part of the style into an inline style in the `render` function.

elijahverdoorn commented on views/components/collapsibleBlock.js:
```diff
@@ -0,0 +1,59 @@
+    }
+  }
+
+  statusColor(){
```
This function is redundant, it's clearer to just use `this.props.borderColor` wherever you need it.

hawkrives commented on views/building-hours/building.js:
```diff
@@ -16,75 +15,106 @@ import * as c from '../components/colors'
+}
+
+let getBorderColor = function(){
+  borderColor = borderColors[openStatus]
```
I don't like global variables. Could you move this discovery process back into the BuildingView function, please?

hawkrives commented on views/components/collapsibleBlock.js:
```diff
@@ -0,0 +1,59 @@
+      },
+      collapsedContent: {
+        alignSelf: 'center',
+        backgroundColor: this.statusColor(),
```
Why don't you just use `this.props.borderColor` here, instead of calling a function?

elijahverdoorn commented on views/building-hours/building.js:
```diff
@@ -16,75 +15,106 @@ import * as c from '../components/colors'
       </Image>
     </View>
+
+    <View style={style().contents}>
```
I think that for consistency we should keep the styling in the `StyleSheet.create()` syntax. (unless there's a good reason to change the entire codebase to use this style that I don't know about!)

elijahverdoorn commented on views/building-hours/building.js:
```diff
@@ -16,75 +15,106 @@ import * as c from '../components/colors'
-  },
-})
+
+let openStatus
```
Why is this defined out here? Can it be defined within the function that it is used in?

elijahverdoorn commented on views/building-hours/building.js:
```diff
@@ -16,75 +15,106 @@ import * as c from '../components/colors'
-  'Open': styles.buildingOpen,
-  'Almost Closed': styles.buildingAlmostClosed,
-  'Closed': styles.buildingClosed,
+  'Open': c.pastelGreen,
```
Just a question: should the color be changed now that it is being used for the background of the text? It looked ok in small amounts on the borders, but I wonder if we want a different shade for the backgrounds. Opinions?

mattk410 replied:
I agree. I changed the color scheme to match the GitHub red/green. I think it looks better with the lighter colors.

hawkrives commented on views/components/collapsibleBlock.js:
```diff
@@ -0,0 +1,59 @@
+
+  static propTypes = {
+    borderColor: React.PropTypes.string.isRequired,
+    children: React.PropTypes.array.isRequired,
```
`children` is better typed as `React.PropTypes.node`, as `this.props.children` is not in an array [if there's only one child](https://facebook.github.io/react/tips/children-props-type.html). `node` covers anything React can render, so elements, text, and arrays of the above.

Commit:
Made changes according to the PR discussion.

-Changed the color scheme to be less bright
-Moved functions/variables to closer scope
-Switched to using StyleSheets

Commit:
Made changes according to the PR discussion.

    -Changed the color scheme to be less bright
    -Moved functions/variables to closer scope
    -Switched to using StyleSheets

hawkrives commented on views/building-hours/all-building-hours.js:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    let dayString = day + ': ' + hoursString
+    dayTimes.push(<Text key={d} style={style}>{dayString}</Text>)
```
Where does `style` come from here?

mattk410 replied:
It's just an argument passed into the function for the purpose of styling the Text. From building.js, it is styles.hoursText.

hawkrives replied:
Ah, ok. I see it now.

hawkrives commented on views/building-hours/all-building-hours.js:
```diff
@@ -0,0 +1,35 @@
+    let day = current.format('dddd')
+    let d = current.format('ddd')
+
+    current = current.add(1, 'days')
```
Why are we adding a day to `current` here when it doesn't appear to be used again?

mattk410 replied:
It is for the next iteration of the loop.

hawkrives replied:
… that makes a lot of sense. This is what I get for doing code reviews at 3am.

hawkrives commented on views/building-hours/all-building-hours.js:
```diff
@@ -0,0 +1,35 @@
+
+    if (timesArray) {
+      let open = moment(timesArray[0], 'H:mm').format('h:mm a')
+      let close = moment(timesArray[1], 'H:mm').format('h:mm a')
```
Don't forget to instantiate these with the final `tz` parameter, too. See https://github.com/StoDevX/AAO-React-Native/blob/master/views/building-hours/get-building-hours.js#L20 for a reference?

hawkrives commented on views/building-hours/all-building-hours.js:
```diff
@@ -0,0 +1,35 @@
+} from 'react-native'
+
+import moment from 'moment-timezone'
+const CENTRAL_TZ = 'America/Winnipeg'
```
:+1:

hawkrives commented on views/building-hours/building.js:
```diff
@@ -6,86 +6,101 @@
+
+  let openStatus = isBuildingOpen(info)
+
+  let styles = StyleSheet.create({
```
I know I said I didn't really like global variables… but stylesheets are kinda an exception? 

Let me try to explain: StyleSheet.create is interesting. Once it's been created, a reference to something like `styles.container` doesn't return an object; it returns a number. That number is a reference into the global stylesheet cache-thing. It's good practice to extract as many styles into a stylesheet as possible, so that they can be created once and just referenced after that. 

Now, that obviously doesn't work if something is dependent on an outside variable – for instance, `borderColor`. That's fine– you just switch to an inline style, but only for the properties that need it.

mattk410 replied:
How do I give StyleSheet.create access to borderColors and openStatus if they are in BuildingView and StyleSheet.create is not?

mattk410 replied:
Never mind

hawkrives replied:
:+1:

hawkrives commented on views/building-hours/building.js:
```diff
@@ -6,86 +6,101 @@
+      borderRightWidth: 5,
+      marginTop: 10,
+      height: 100,
+      borderColor: borderColors[openStatus],
```
So everything _but_ this line would be extracted into a global `styles` object in this file.

hawkrives commented on views/building-hours/building.js:
```diff
@@ -6,86 +6,101 @@
+    contents: {
+      paddingBottom: 10,
+      paddingTop: 5,
+      width: Dimensions.get('window').width,
```
Oh, and excluding this, too, since it needs to be able to respond to size changes.

hawkrives commented on views/building-hours/building.js:
```diff
@@ -6,86 +6,101 @@
-    </View>
-  )
+    <CollapsibleBlock style={styles} backgroundColor={borderColors[openStatus]}>
+      <View style={[styles.container]}>
```
This would become `style={[styles.container, {borderColor: borderColors[openStatus]}]}` – so that the borderColor is the only thing set by the inline style.

hawkrives commented on views/building-hours/building.js:
```diff
@@ -6,86 +6,101 @@
+        </Image>
+      </View>
+
+      <View style={styles.contents}>
```
This would become `style={[styles.container, {width: Dimensions.get('window').width}]}`

hawkrives commented on views/building-hours/building.js:
```diff
@@ -6,86 +6,101 @@
+    'Closed': '#F7C8C8',
+  }
+
+  type PropsType = {
```
This `type` declaration should live outside of the function.

hawkrives commented on views/building-hours/index.js:
```diff
@@ -5,7 +5,7 @@
 
 import React from 'react'
-import {View, Text, ListView, RefreshControl} from 'react-native'
+import {ListView, RefreshControl} from 'react-native'
```
Thanks

hawkrives commented on views/components/collapsibleBlock.js:
```diff
@@ -0,0 +1,53 @@
+export default class CollapsibleBlock extends React.Component {
+
+  static propTypes = {
+    backgroundColor: React.PropTypes.string.isRequired,
```
Here, a more general approach might be nice – what do you think about accepting a `style` property, instead of just a single `backgroundColor` prop? If you don't want to, that's fine – it's unlikely that we'll need to pass more styles down, but it might be nice ð

hawkrives commented on views/components/collapsibleBlock.js:
```diff
@@ -0,0 +1,53 @@
+        <TouchableOpacity
+          onPress={() => this.setState({isCollapsed: !this.state.isCollapsed})}
+        >
+          {this.props.children[0]}
```
Um… Let me think about this and get back to you. It's generally a bad idea to do anything to `props.children`, as it won't be an array if there's only one child… which there shouldn't be, but … hmm. Referencing https://facebook.github.io/react/tips/children-props-type.html…

It looks like we should use the React.children.toArray helper function to ensure that this is an array: https://facebook.github.io/react/docs/top-level-api.html#react.children.toarray

mattk410 replied:
There should be exactly 2 children in the array. One is the TouchableOpacity "header" and the other is the dropdown contents

hawkrives commented on views/components/collapsibleBlock.js:
```diff
@@ -0,0 +1,53 @@
+          onPress={() => this.setState({isCollapsed: !this.state.isCollapsed})}
+        >
+          {this.props.children[0]}
+          <Collapsible collapsed={this.state.isCollapsed} style={[this.styles.collapsedContent, {backgroundColor: this.props.backgroundColor}]}>
```
Anyway, if you wanted to make the styles more general, you could just replace this with `style={[styles.collapsedContent, this.props.style]}`.

hawkrives commented on views/components/collapsibleBlock.js:
```diff
@@ -0,0 +1,53 @@
+    isCollapsed: true,
+  }
+
+  styles = StyleSheet.create({
```
Please remove this from the class?

Commit:
Made changes based on Hawken's suggestions.

-Moved StyleSheets outside of functions
-Pass styles into Collapsible component

Commit:
Merge branch 'master' into revamp-building-hours
